Round decimal record difference percent to three decimals after zeros

mattfergoda opened the pull request:
<!-- prettier-ignore -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #1581 by @AetherUnbound 

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

Round the data refresh percent change report to 3 digits after the decimal, ignoring zeros. E.g. 100.012365 -> 100.0124.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
`just catalog/dags/test`


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

sarayourfriend commented on catalog/dags/data_refresh/reporting.py:
```diff
@@ -30,9 +30,29 @@ def report_record_difference(before: str, after: str, media_type: str, dag_id: s
     return message
+
+
+def _round_sig_figs(num, n):
+    """
+    Round number to the nearest n digits.
+    If num is an integer, don't change it.
+    If num has a decimal, round to 3 significant figures (after any zeros),
+    rounding up if the next digit is 5 or greater.
+    """
+
+    if num % 1 == 0:
+        return int(num)
+
+    # Round decimal part to 3 sig figs. Leave int part untouched.
+    [int_part, dec_part] = str(num).split(".")
+
+    # Have to awkwardly prepend the decimal part with '0.' to get correct
+    # rounding behavior, then slice '0.' off before concatenating
+    # the final number.
+    return float(int_part + "." + "{:.{n}g}".format(float("0." + dec_part), n=n)[2:])
```
This is over complicating things a bit, I think for the sake of always having 3 significant digits. Using some fancy [string format spec](https://docs.python.org/3/library/string.html#format-specification-mini-language), the whole thing can be simplified, so long as we don't care that `0.00005646%` will get truncated to just `0.00006%`... and frankly, at such small percentages, I really don't think it matters to have accuracy past 1/100, it just isn't interesting or useful information to have such abstract and small percentages.

I've left a suggestion on the line above for format spec to use.

AetherUnbound replied:
@sarayourfriend I was talking about this with @mattfergoda in person - many times for the data refresh, we do have changes that are quite small by percentage but still register as a +/- half a million record difference. Things like `0.0654%` or even `0.00481%`. For me, and perhaps this is just my own perspective, truncating those down completely does lose some of the specificity we have. Maybe it doesn't matter, but whatever change we make I'd like to _at least_ be sure we have a few digits available a couple of decimal points out.

sarayourfriend replied:
What do you mean by "truncating them down completely"? As I wrote:

> 0.00005646% will get truncated to just 0.00006%

In what way is this a meaningful difference in the information, when the actual count is also present?

AetherUnbound replied:
Those are all fair points! We can go with the much simpler implementation with that in mind 🙂

sarayourfriend commented on catalog/dags/data_refresh/reporting.py:
```diff
@@ -30,9 +30,29 @@ def report_record_difference(before: str, after: str, media_type: str, dag_id: s
 _Note: All values are row estimates and are not (but nearly) exact_
 *Record count difference for `{media_type}`*: {before:,} → {after:,}
-*Change*: {count_diff:+,} ({percent_diff:+}% Δ)
+*Change*: {count_diff:+,} ({_round_sig_figs(percent_diff, 3):+}% Δ)
```
As mentioned below, the `_round_sig_figs` can be removed if we simplify things using just format spec significant digits instructions.

If we use `g`, the general format, this will use scientific notation in some cases:

```
>>> "{n:+3g}".format(n=0.0000512346888129371512345)
'+5.12347e-05'
```

So, I think it would be better to remove the `* 100` on line 24, and use the `%` spec instead:

```
>>> "{n:+3%}".format(n=0.000000512346888129371512345)
'+0.000051%'
```

```suggestion
*Change*: {count_diff:+,} (percent_diff:+3%} Δ)
```

If you apply that suggestion, please update line 24 as well.

AetherUnbound replied:
FWIW the `3` here does not seem to have any effect 🤔 
```python
[nav] In [3]: "{n:+3%}".format(n=0.001000512346888129371512345)
Out[3]: '+0.100051%'

[nav] In [4]: "{n:+3%}".format(n=0.001000512340888129371512345)
Out[4]: '+0.100051%'

[nav] In [5]: "{n:+3%}".format(n=0.0000512340888129371512345)
Out[5]: '+0.005123%'

[nav] In [6]: "{n:+2%}".format(n=0.0000512340888129371512345)
Out[6]: '+0.005123%'

[nav] In [7]: "{n:+1%}".format(n=0.0000512340888129371512345)
Out[7]: '+0.005123%'

[nav] In [8]: "{n:+%}".format(n=0.0000512340888129371512345)
Out[8]: '+0.005123%'

```

mattfergoda replied:
If `"{n:+3%}".format(num)` is the agreed way forward, I'm happy to make this change and update the tests.

sarayourfriend replied:
The number is apparently ignored for `%`, TIL. It does have an effect on general formatting, and I converted to `%` from that, not realising it wouldn't make a difference.

```
>>> "{n:+3g}".format(n=52.212331223456234512)
'+52.2123'
>>> "{n:+3g}".format(n=0.001000512346888129371512345)
'+0.00100051'
>>> "{n:+3g}".format(n=0.6888129371512345)
'+0.688813'
>>> "{n:+3g}".format(n=0.000000000000000000000000888129371512345)
'+8.88129e-25'
```

AetherUnbound replied:
Sara's right, I think this approach will be way more straightforward than anything I'm suggesting. If you're comfortable with making those changes Matt, please proceed!

sarayourfriend reviewed (changes requested):
Thanks for the PR @mattfergoda! I'm pretty sure there is a simpler way to accomplish this, as long as we don't get too caught up with always wanting three digits on very small decimals. For our context, I'd say that level of precision is not worth the code required to achieve it, when using format-spec will give us all the information we could possibly actually need and use.

mattfergoda commented:
@AetherUnbound @sarayourfriend I simplified the rounding logic based on the above conversation and updated the tests accordingly. I added a parametrized test case to the existing ones where the decimal value is truncated.

sarayourfriend reviewed (commented):
@mattfergoda, thanks for updating the PR. Left one comment, but it's unclear to me whether a change is needed until CI runs. Can you rebase the PR for that?

AetherUnbound reviewed (approved):
Thank you for the contribution @mattfergoda, and for your patience while iterating on this issue! It looks like GitHub is having an outage at the moment which is preventing CI from running, but this LGTM.

AetherUnbound commented:
Ah, it looks like this is conflicting with #4067 - it seems easiest to squash the commits and rebase, in the interest of time I'll go ahead and take care of that.

sarayourfriend reviewed (approved):
LGMT! Thanks again @mattfergoda :+1:
